### PR TITLE
CAS authentication.

### PR DESCRIPTION
**Pull request by nrmay**
This change adds support for the JASIG Central Authentication Service, including:
* package: tardis.tardis_portal.auth.cas package
* migration: 0005
* additional tests
* documentation of settings added to authorisation.rst

See http://jasig.github.io/cas for further information about CAS.

### CODE REVIEW

**Comment by nrmay** [comment]
Note: this change has been tested on:
* Ubuntu 14.04
* CentOS 7

**Comment by nrmay** [comment]
Why do I keep hitting the 'Close and Comment' button?

File: tardis/tardis_portal/auth/cas/models.py [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+from exceptions import CasTicketException, CasConfigException
+# Ed Crewe - add in signals to delete old tickets
+from django.db.models.signals import post_save
+from datetime import datetime, timedelta
+
+from tardis.tardis_portal.auth.cas import CAS
+
+class Tgt(models.Model):
+    username = models.CharField(max_length = 255, unique = True)
+    tgt = models.CharField(max_length = 255)
```
Comment (grischa) [inline tardis/tardis_portal/auth/cas/models.py]: Codacy Issue found: [No space allowed around keyword argument assignment](https://www.codacy.com/app/mytardis/mytardis/file/1211211486/issues/source?bid=2125231&fileBranchId=2928880#l16)

File: tardis/tardis_portal/auth/cas/models.py [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+    pgtIou = models.CharField(max_length = 255, unique = True)
+    tgt = models.CharField(max_length = 255)
+    created = models.DateTimeField(auto_now = True)
+
+def get_tgt_for(user):
+    if not settings.CAS_PROXY_CALLBACK:
+        raise CasConfigException("No proxy callback set in settings")
+
+    try:
+        return Tgt.objects.get(username = user.username)
```
Comment (grischa) [inline tardis/tardis_portal/auth/cas/models.py]: Codacy Issue found: [No space allowed around keyword argument assignment](https://www.codacy.com/app/mytardis/mytardis/file/1211211486/issues/source?bid=2125231&fileBranchId=2928880#l60)

File: tardis/tardis_portal/auth/cas/backends.py [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+
+    if tree.find(CAS + 'authenticationSuccess') is not None:
+        username = tree.find(CAS + 'authenticationSuccess/' + CAS + 'user').text
+        pgtIouIdElement = tree.find(CAS + 'authenticationSuccess/' + CAS + 'proxyGrantingTicket')
+        pgtIouId = pgtIouIdElement.text if pgtIouIdElement is not None else None
+
+        if pgtIouId:
+            pgtIou = PgtIOU.objects.get(pgtIou = pgtIouId)
+            try:
+                tgt = Tgt.objects.get(username = username)
```
Comment (grischa) [inline tardis/tardis_portal/auth/cas/backends.py]: Codacy Issue found: [No space allowed around keyword argument assignment](https://www.codacy.com/app/mytardis/mytardis/file/1211211482/issues/source?bid=2125231&fileBranchId=2928880#l75)

File: tardis/tardis_portal/auth/cas/views.py [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+    if not next_page:
+        next_page = _redirect_url(request)
+    if request.user.is_authenticated():
+        #message = "You are logged in as %s." % request.user.username
+        #request.user.message_set.create(message=message)
+        return HttpResponseRedirect(next_page)
+    ticket = request.GET.get('ticket')
+    service = _service_url(request, next_page)
+    if ticket:
+        logger.debug("auth.authenticate(ticket[%s], service[%s])" % (ticket,service))
```
Comment (grischa) [inline tardis/tardis_portal/auth/cas/views.py]: Codacy Issue found: [Exactly one space required after comma](https://www.codacy.com/app/mytardis/mytardis/file/1211211488/issues/source?bid=2125231&fileBranchId=2928880#l114)

File: tardis/tardis_portal/auth/cas/views.py [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+    logger.debug("request[%s]" % (request))
+
+    pgtIou = request.GET.get('pgtIou')
+    tgt = request.GET.get('pgtId')
+
+    if not (pgtIou and tgt):
+        return HttpResponse()
+
+    try:
+        PgtIOU.objects.create(tgt = tgt, pgtIou = pgtIou, created = datetime.now())
```
Comment (grischa) [inline tardis/tardis_portal/auth/cas/views.py]: Codacy Issue found: [No space allowed around keyword argument assignment](https://www.codacy.com/app/mytardis/mytardis/file/1211211488/issues/source?bid=2125231&fileBranchId=2928880#l167)

File: tardis/tardis_portal/auth/cas/backends.py [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+    tree = ElementTree.fromstring(response)
+    page.close()
+
+    if tree.find(CAS + 'authenticationSuccess') is not None:
+        username = tree.find(CAS + 'authenticationSuccess/' + CAS + 'user').text
+        pgtIouIdElement = tree.find(CAS + 'authenticationSuccess/' + CAS + 'proxyGrantingTicket')
+        pgtIouId = pgtIouIdElement.text if pgtIouIdElement is not None else None
+
+        if pgtIouId:
+            pgtIou = PgtIOU.objects.get(pgtIou = pgtIouId)
```
Comment (grischa) [inline tardis/tardis_portal/auth/cas/backends.py]: Codacy Issue found: [No space allowed around keyword argument assignment](https://www.codacy.com/app/mytardis/mytardis/file/1211211482/issues/source?bid=2125231&fileBranchId=2928880#l73)

File: tardis/tardis_portal/auth/cas/views.py [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+    logger.debug("service[%s] ticket[%s]" % (service, ticket))
+
+    LOGINS = {'ST':'login',
+              'PT':'proxyValidate'}
+    params = {'service': service}
+    if settings.CAS_EXTRA_LOGIN_PARAMS:
+        params.update(settings.CAS_EXTRA_LOGIN_PARAMS)
+    if not ticket:
+        ticket = 'ST'
+    login = LOGINS.get(ticket[:2],'login')
```
Comment (grischa) [inline tardis/tardis_portal/auth/cas/views.py]: Codacy Issue found: [Exactly one space required after comma](https://www.codacy.com/app/mytardis/mytardis/file/1211211488/issues/source?bid=2125231&fileBranchId=2928880#l81)

File: tardis/tardis_portal/auth/cas/models.py [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+            if tree.find(CAS + 'proxySuccess') is not None:
+                return tree.find(CAS + 'proxySuccess/' + CAS + 'proxyTicket').text
+            else:
+                raise CasTicketException("Failed to get proxy ticket")
+        finally:
+            page.close()
+
+class PgtIOU(models.Model):
+    """ Proxy granting ticket and IOU """
+    pgtIou = models.CharField(max_length = 255, unique = True)
```
Comment (grischa) [inline tardis/tardis_portal/auth/cas/models.py]: Codacy Issue found: [No space allowed around keyword argument assignment](https://www.codacy.com/app/mytardis/mytardis/file/1211211486/issues/source?bid=2125231&fileBranchId=2928880#l51)

File: tardis/tardis_portal/auth/cas/models.py [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+from django.core.exceptions import ObjectDoesNotExist
+from exceptions import CasTicketException, CasConfigException
+# Ed Crewe - add in signals to delete old tickets
+from django.db.models.signals import post_save
+from datetime import datetime, timedelta
+
+from tardis.tardis_portal.auth.cas import CAS
+
+class Tgt(models.Model):
+    username = models.CharField(max_length = 255, unique = True)
```
Comment (grischa) [inline tardis/tardis_portal/auth/cas/models.py]: Codacy Issue found: [No space allowed around keyword argument assignment](https://www.codacy.com/app/mytardis/mytardis/file/1211211486/issues/source?bid=2125231&fileBranchId=2928880#l15)

File: tardis/tardis_portal/auth/cas/models.py [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+    tgt = models.CharField(max_length = 255)
+    created = models.DateTimeField(auto_now = True)
```
Comment (grischa) [inline tardis/tardis_portal/auth/cas/models.py]: Codacy Issue found: [No space allowed around keyword argument assignment](https://www.codacy.com/app/mytardis/mytardis/file/1211211486/issues/source?bid=2125231&fileBranchId=2928880#l17)

**Comment by nrmay** [comment]
Closed for admin reasons.

**Pull request closed without merging**